kbfsedits/data_types.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package kbfsedits
+
+import (
+	"time"
+
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/kbfscrypto"
+	"github.com/keybase/kbfs/kbfsmd"
+	"github.com/keybase/kbfs/tlf"
+)
+
+// NotificationVersion is the numeric version of the edit notification
+// data structure.
+type NotificationVersion int
+
+const (
+	// NotificationV1 is unused, but represents the original,
+	// MD-ops-based edit notification list.
+	NotificationV1 NotificationVersion = 1
+	// NotificationV2 is the first version that stores JSON-encoded
+	// notifications in chat messages.
+	NotificationV2 NotificationVersion = 2
+)
+
+// NotificationOpType indicates the type of the operation that caused
+// the notification.
+type NotificationOpType string
+
+const (
+	// NotificationCreate is the type of an edit notification
+	// representing a new file or directory creation.
+	NotificationCreate NotificationOpType = "create"
+	// NotificationModify is the type of an edit notification
+	// representing a file modification.
+	NotificationModify NotificationOpType = "modify"
+	// NotificationRename is the type of an edit notification
+	// representing a file or directory getting renamed.
+	NotificationRename NotificationOpType = "rename"
+	// NotificationDelete is the type of an edit notification
+	// representing a file or directory getting deleted.
+	NotificationDelete NotificationOpType = "delete"
+)
+
+// EntryType indicates the type of the file that was edited.
+type EntryType string
+
+const (
+	// EntryTypeFile is for files that have been edited.  Note that
+	// covers both regular files and executables.
+	EntryTypeFile EntryType = "file"
+	// EntryTypeDir is for directories that have been edited.
+	EntryTypeDir EntryType = "dir"
+	// EntryTypeSym is for symlinks that have been edited.
+	EntryTypeSym EntryType = "sym"
+)
+
+// NotificationParams is used for op-type-specific data.
+type NotificationParams struct {
+	OldFileName string `json:",omitempty"` // for renames
+	Offset      uint64 `json:",omitempty"` // for modifies
+	Length      uint64 `json:",omitempty"` // for modifies
+}
```

Comment:
If I copy a file into KBFS and override an existing file, would that be an array of modifies, first one being truncate and following being writes?

Reply:
Good question.  Internally, in the MD objects, that's described with a `createOp` operation that includes unreferenced blocks, indicating that a file was overwritten.  In this new protocol, we could change that potentially to show a delete followed by a create, to make it more explicit.

Reply:
I see. Would there still be a bunch of "modifies" in this case? It seems it'd be nice if we could buffer them and collapse notifications. (Not in scope of this PR though.)

Reply:
The write batching system already collapses them into ~1 second windows, so at first I was just planning to mirror that.  Doing anything else will require editing the chat history, or delaying sends for a while to batch stuff together.

The list we'll be returning to the frontend could be have more-collapsed ranges, but actually I think the frontend list will be way more collapsed, and will probably only contain filenames and timestamps. I could be wrong though.